### ISSUE TITLE
Being able to apply all custom filters on Due and Overdue visits cards added to custom dashboard

### ISSUE THREAD

**Issue by mahalakshme** [avniproject/avni-client#1382]
### Need:
- When specific filters for subject type like mentioned [here](https://github.com/avniproject/avni-webapp/issues/1204) it needs to work in mobile app
Example usecase: To support 'Weight for Age grade' in JSSCP that is configured currently in MyDashboard for a specific subject type.
- When the same filter types(Gender, Registeration Date, Enrolment Date, Encounter Date, Program Encounter Date, Address, Concept, Group Subject) are configured generically without being under 'Subject Type' filter as well it need to work
Example usecase: To support filtering by a concept present in registeration forms of all subject types
![Image](https://github.com/avniproject/avni-client/assets/10369963/c737121c-0cc2-4e76-b6c8-73e103eb0537)

### As is:
Currently only Location filter is supported on Due and Overdue visits card when configured generically(without being based on subject type)

### AC:
- All the above mentioned filters when configured(with or without being specific to subject type) need to work when 'Due/Scheduled visits' and 'Overdue visits' card added to custom dashboard.

### Technical details:
- Made applying of filters with and w/o subject type as part of same card, since its good to have the same code for both cases and the subject type filter will only get applied additionally in the 1st case.

**Comment by himeshr** [comment]
Code review and QA is pending for below specified cards, only after which we should resume further deveopment here as all changes in these cards have been throwing up errors or blocking dev org configuration.

avniproject/avni-webapp#1203
avniproject/avni-webapp#1204
avniproject/avni-webapp#1235
avniproject/avni-client#1381

**Comment by petmongrels** [comment]
### Card dev notes
- check if filter cache is working when there are more than one gender filter values (done)
- removing subject types should remove the program/encounter types as well (done)
- test with multiple address levels on user's device (done)
- throw error to check whether cache is getting reset (done)

#### Design doc topics
- How checksum is maintained
- How and why handle errors during filtering
- Responsibility definition

#### QA notes
- Regression required for MyDashboard (especially address levels)
- Since concept has multiple data types, testing should be done for this card for different data types
- Please also check whether each type of filters are getting remembered
- Also check any implementation that have used Custom Report cards based on rules recently, to check if they are working fine too.

**Comment by mahalakshme** [comment]
@petmongrels getting this error after configuring the below dashboard. Seems to be getting after deleting the dashboard as well



![Image](https://github.com/avniproject/avni-client/assets/10369963/19e5adbf-c79a-46b0-8a2c-15959ba8e8c1)

![Image](https://github.com/avniproject/avni-client/assets/10369963/5329bdbb-6900-4729-a8d8-a949175df674)

**Comment by AchalaBelokar** [comment]
Env: prereleases 
- [ ] Login with maha@apfodisha and created a card household with group subject type filter . but filter is not working.


https://github.com/user-attachments/assets/56bd688c-607b-4210-8606-a6f68972674d

**Comment by petmongrels** [comment]
fixed. the second report card will show the correct numbers now.
FYI - Please note that the first report card will show zero because it is showing number of households which is part of group household. not a valid condition.

**Comment by AchalaBelokar** [comment]
- [ ] I login with maha@apfodisha and go to the household dashboard and added the location filter but it is not showing the value it is giving me error
Cannot set property 'isSelected' of undefined
at anonymous (address at index.android.bundle:1:2417179)
at forEach ((native)::)
at setSelectedAddresses (address at index.android.bundle:1:2417114)
at anonymous (address at index.android.bundle:1:4248687)
at anonymous (address at index.android.bundle:1:1112823)
at map ((native)::)
at apply ((native)::)
at value (address at index.android.bundle:1:2247068)
at map (address at index.android.bundle:1:1112785)
at render (address at index.android.bundle:1:4247660)
at finishClassComponent (address at index.android.bundle:1:367544)
at updateClassComponent (address at index.android.bundle:1:367405)
at beginWork$1 (address at index.android.bundle:1:399723)
at performUnitOfWork (address at index.android.bundle:1:386774)
at workLoopSync (address at index.android.bundle:1:386676)
at renderRootSync (address at index.android.bundle:1:386560)
at performSyncWorkOnRoot (address at index.android.bundle:1:384043)
at flushSyncCallbacks (address at index.android.bundle:1:349073)
at batchedUpdatesImpl (address at index.android.bundle:1:400497)
at batchedUpdates (address at index.android.bundle:1:341930)
at _receiveRootNodeIDEvent (address at index.android.bundle:1:342203)
at receiveTouches (address at index.android.bundle:1:395353)
at apply ((native)::)
at __callFunction (address at index.android.bundle:1:98035)
at anonymous (address at index.android.bundle:1:96528)
at __guard (address at index.android.bundle:1:97468)
at callFunctionReturnFlushedQueue (address at index.android.bundle:1:96486)

https://github.com/user-attachments/assets/d64420e7-e319-414c-8901-72e36f525ecd

**Comment by petmongrels** [comment]
moving it to QA, as the prerelease environment has been refreshed

**Comment by AchalaBelokar** [comment]
- [ ]I created the new card named household and the card is not displayed in the mobile app 
https://drive.google.com/file/d/1PNGGJmm_GqsTRhmq0Iow1WeoHOqFL5FG/view?usp=sharing

**Comment by himeshr** [comment]
### Details
Env: prerelease
Org: apf
User: dineshg

Related links:
https://prerelease.avniproject.org/#/appDesigner/dashboard/616/show

### Issue description
Program Encounter date filters seem to be incorrectly applied for Overdue visits as seen in following screen-recording

https://drive.google.com/file/d/1FJQ2YiOCwpZN5ZUeDg6cRz371v2k9c-_/view?usp=sharing

**Comment by himeshr** [comment]
> ### Details
> Env: prerelease Org: apf User: dineshg
> 
> Related links: https://prerelease.avniproject.org/#/appDesigner/dashboard/616/show
> 
> ### Issue description
> Program Encounter date filters seem to be incorrectly applied for Overdue visits as seen in following screen-recording
> 
> https://drive.google.com/file/d/1FJQ2YiOCwpZN5ZUeDg6cRz371v2k9c-_/view?usp=sharing

Based on discussions with Maha, this seems to be current behaviour on mydashboard, where-in the progEncDate filter is applied on all encounters of individual and then we search for individuals with due visits amongst them.. Instead of searching for dueVisits with scheduledDate for the ProgEncDate filter value.

**Comment by himeshr** [comment]
### THIS HAS TO BE FIXED

MyDashboard filter by registrationDate is broken.. Continue testing with Prod old apk to confirm that this was already the case earlier.

Comparision btwn APP versions 10.0 and 9.4 shows difference in results for the same user and org..
<img width="1728" alt="Screenshot 2024-09-20 at 12 10 27 PM" src="https://github.com/user-attachments/assets/9911060e-7774-42d8-b987-e5c823024ce8">

**Comment by himeshr** [comment]
### THIS IS NOT TO BE FIXED

Issues found with new(10.0) and old (9.4) version of Avni APK MyDashboard are as follows:
- MyDashboard, Individual Registration Date filter with value causes issues on relaunch of Avni app, Count of Due and Overdue visits cards are wrong
- MyDashboard, ProgEncounterType filter values are not set correctly on Filters Screen, though they show up in applied filters section

https://drive.google.com/file/d/1Oif2BInTycQVx-xEGGYM9rwdcsOfyBGs/view?usp=sharing

**Comment by petmongrels** [comment]
### not fixed another issue noticed
- registration date range filter applied
- restart app
- filter is shown as applied but the counts are not calculated as per the filter

Same issues exists in 9.4 app

**Comment by himeshr** [comment]
MyDashboard filter by registrationDate is now fixed.. Did not find any new issues that were not present in older apk versions.
